File: PYTHON/contexa/workspace.py
"""
GCC Workspace — core file-system layer.
Based on arXiv:2508.00031v2.

The paper specifies a .GCC/ root directory containing:
  main.md              — global roadmap / planning artifact
  branches/
    <branch_name>/
      log.md           — continuous OTA (Observation-Thought-Action) trace
      commit.md        — milestone-level commit summaries
      metadata.yaml    — branch intent, status, creation info

All four commands (COMMIT, BRANCH, MERGE, CONTEXT) operate on this structure.
"""

import os
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from .models import (
    BranchMetadata,
    CommitRecord,
    ContextResult,
    OTARecord,
)


MAIN_BRANCH = "main"
GCC_DIR = ".GCC"


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _short_id() -> str:
    return str(uuid.uuid4())[:8]


class GCCWorkspace:
    """
    Manages the .GCC/ directory structure for one agent project.

    Usage
    -----
    ws = GCCWorkspace("/path/to/project")
    ws.init("Build a REST API service")

    # agent logs OTA steps
    ws.log_ota("file list returned", "need to read main.py", "read_file main.py")

    # agent commits a milestone
    ws.commit("Implement GET /users endpoint")

    # agent branches to explore alternative
    ws.branch("auth-jwt", "Explore JWT-based auth vs session auth")

    # agent merges branch back
    ws.merge("auth-jwt")

    # retrieve context (K=1 by default, per paper experiments)
    ctx = ws.context(k=1)
    print(ctx.summary())
    """

    def __init__(self, project_root: str):
        self.root = Path(project_root)
        self.gcc_dir = self.root / GCC_DIR
        self._current_branch: str = MAIN_BRANCH

    def _branch_dir(self, branch: str) -> Path:
        return self.gcc_dir / "branches" / branch

    def _log_path(self, branch: str) -> Path:
        return self._branch_dir(branch) / "log.md"

    def _commit_path(self, branch: str) -> Path:
        return self._branch_dir(branch) / "commit.md"

    def _meta_path(self, branch: str) -> Path:
        return self._branch_dir(branch) / "metadata.yaml"

    def _main_md(self) -> Path:
        return self.gcc_dir / "main.md"

    def _read(self, path: Path) -> str:
        if path.exists():
            return path.read_text(encoding="utf-8")
        return ""

    def _append(self, path: Path, text: str) -> None:
        with open(path, "a", encoding="utf-8") as f:
            f.write(text)

    def _write(self, path: Path, text: str) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(text, encoding="utf-8")

    def _read_commits(self, branch: str) -> list[CommitRecord]:
        """Parse commit.md into a list of CommitRecord objects."""
        text = self._read(self._commit_path(branch))
        records: list[CommitRecord] = []
        if not text:
            return records
        for block in text.split("---\n"):
            block = block.strip()
            if not block:
                continue
            lines = block.splitlines()
            commit_id = branch_purpose = prev_summary = contribution = ts = ""
            for line in lines:
                if line.startswith("## Commit"):
                    commit_id = line.split("`")[1] if "`" in line else ""
                elif line.startswith("**Timestamp:**"):
                    ts = line.split("**Timestamp:**")[-1].strip()
                elif line.startswith("**Branch Purpose:**"):
                    branch_purpose = line.split("**Branch Purpose:**")[-1].strip()
                elif line.startswith("**Previous Progress Summary:**"):
                    prev_summary = line.split("**Previous Progress Summary:**")[
                        -1
                    ].strip()
                elif line.startswith("**This Commit's Contribution:**"):
                    contribution = line.split("**This Commit's Contribution:**")[
                        -1
                    ].strip()

            if commit_id:
                records.append(
                    CommitRecord(
                        commit_id=commit_id,
                        branch_name=branch,
                        branch_purpose=branch_purpose,
                        previous_progress_summary=prev_summary,
                        this_commit_contribution=contribution,
                        timestamp=ts,
                    )
                )

        return records

    def _read_ota(self, branch: str) -> list[OTARecord]:
        text = self._read(self._log_path(branch))
        records: list[OTARecord] = []
        if not text:
            return records
        for block in text.split("--------\n"):
            block = block.strip()
            if not block:
                continue
            lines = block.splitlines()
            step = 0
            ts = obs = thought = action = ""
            for line in lines:
                if line.startswith("### Step"):
                    header = line.replace("### Step", "").strip()
                    parts = header.split("-", 1)
                    try:
                        step = int(parts[0].strip())
                    except ValueError:
                        step = 0
                    ts = parts[1].strip() if len(parts) > 1 else ""
                elif line.startswith("**Observation:**"):
                    obs = line.split("**Observation:**")[-1].strip()
                elif line.startswith("**Thought:**"):
                    thought = line.split("**Thought:**")[-1].strip()
                elif line.startswith("**Action:**"):
                    action = line.split("**Action:**")[-1].strip()
            if obs or thought or action:
                records.append(OTARecord(ts, obs, thought, action, step))

        return records

    def init(self, project_roadmap: str = "") -> None:
        """
        Initialise a new GCC workspace. Creates .GCC/ structure with a
        'main' branch, main.md roadmap, and initial metadata.
        """
        if self.gcc_dir.exists():
            raise FileExistsError(f"GCC workspace already exists at {self.gcc_dir}")
        main_branch_dir = self._branch_dir(MAIN_BRANCH)
        main_branch_dir.mkdir(parents=True)
        roadmap_content = (
            f"# Project Roadmap\n\n**Initialized:** {_now()}\n\n{project_roadmap}\n"
        )
        self._write(self._main_md(), roadmap_content)
        self._write(
            self._log_path(MAIN_BRANCH), f"# OTA Log — branch `{MAIN_BRANCH}`\n\n"
        )
        self._write(
            self._commit_path(MAIN_BRANCH),
            f"# Commit History — branch `{MAIN_BRANCH}`\n\n",
        )

        meta = BranchMetadata(
            name=MAIN_BRANCH,
            purpose="Primary reasoning trajectory",
            created_from="",
            created_at=_now(),
            status="active",
        )
        self._write(self._meta_path(MAIN_BRANCH), meta.to_yaml())

        self._current_branch = MAIN_BRANCH

    def load(self) -> None:
        """Load an existing GCC workspace (no-op if already in memory)."""
        if not self.gcc_dir.exists():
            raise FileNotFoundError(f"No GCC workspace found at {self.gcc_dir}")
        # Determine current branch from active metadata
        self._current_branch = MAIN_BRANCH

    def log_ota(self, observation: str, thought: str, action: str) -> OTARecord:
        """
        Append an OTA step to the current branch's log.md.
        The paper logs continuous OTA cycles: Observation-Thought-Action.
        """
        existing = self._read_ota(self._current_branch)
        step = len(existing) + 1
        record = OTARecord(
            timestamp=_now(),
            observation=observation,
            thought=thought,
            action=action,
            step=step,
        )
        self._append(self._log_path(self._current_branch), record.to_markdown())
        return record

    def commit(
        self,
        contribution: str,
        previous_summary: Optional[str] = None,
        update_roadmap: Optional[str] = None,
    ) -> CommitRecord:
        """
        COMMIT command (paper S3.2).

        Persists a milestone checkpoint to commit.md. Fields per paper:
          - Branch Purpose (from metadata)
          - Previous Progress Summary
          - This Commit's Contribution

        Optionally updates main.md global roadmap.
        """
        meta_text = self._read(self._meta_path(self._current_branch))
        meta = BranchMetadata.from_yaml(meta_text) if meta_text else None
        branch_purpose = meta.purpose if meta else ""

        # Auto-summarise previous progress from last commit if not provided
        if previous_summary is None:
            commits = self._read_commits(self._current_branch)
            if commits:
                previous_summary = commits[-1].this_commit_contribution
            else:
                previous_summary = "Initial state — no prior commits."

        record = CommitRecord(
            commit_id=_short_id(),
            branch_name=self._current_branch,
            branch_purpose=branch_purpose,
            previous_progress_summary=previous_summary,
            this_commit_contribution=contribution,
            timestamp=_now(),
        )
        self._append(self._commit_path(self._current_branch), record.to_markdown())

        # Optionally update the global roadmap in main.md
        if update_roadmap:
            self._append(
                self._main_md(),
                f"\n## Update ({record.timestamp})\n{update_roadmap}\n",
            )

        return record

    def branch(self, name: str, purpose: str) -> "GCCWorkspace":
        """
        BRANCH command (paper S3.3).

        Creates an isolated workspace for alternative plans or experimental
        reasoning. Mathematically: B_t^(name) = BRANCH(M_{t-1}).

        Initialises empty OTA trace (log.md) and commit.md for the branch,
        carrying forward the parent's last committed state as starting context.
        """
        branch_dir = self._branch_dir(name)
        if branch_dir.exists():
            raise FileExistsError(f"Branch '{name}' already exists.")

        branch_dir.mkdir(parents=True)

        # Initialise empty OTA log (fresh execution trace)
        self._write(self._log_path(name), f"# OTA Log — branch `{name}`\n\n")
        self._write(self._commit_path(name), f"# Commit History — branch `{name}`\n\n")

        # metadata.yaml records intent and motivation (paper S3.3)
        meta = BranchMetadata(
            name=name,
            purpose=purpose,
            created_from=self._current_branch,
            created_at=_now(),
            status="active",
        )
        self._write(self._meta_path(name), meta.to_yaml())

        # Switch current branch
        self._current_branch = name
        return self

    def merge(
        self,
        branch_name: str,
        summary: Optional[str] = None,
        target: str = MAIN_BRANCH,
    ) -> CommitRecord:
        """
        MERGE command (paper S3.4).

        Integrates a completed branch back into the target (default: main),
        merging summaries and execution traces to produce a unified,
        consistent state. Creates a merge commit on the target branch.
        """
        # Gather branch commits and synthesise
        branch_commits = self._read_commits(branch_name)
        branch_ota = self._read_ota(branch_name)
        meta_text = self._read(self._meta_path(branch_name))
        meta = BranchMetadata.from_yaml(meta_text) if meta_text else None

        if not summary:
            contributions = " | ".join(
                c.this_commit_contribution for c in branch_commits
            )
            summary = f"Merged branch `{branch_name}` ({len(branch_commits)} commits). Contributions: {contributions}"

        # Append branch OTA to target log (merge trace)
        if branch_ota:
            merge_header = f"\n## Merged from `{branch_name}` ({_now()})\n\n"
            self._append(self._log_path(target), merge_header)
            for rec in branch_ota:
                self._append(self._log_path(target), rec.to_markdown())

        # Create merge commit on target
        self._current_branch = target
        merge_commit = self.commit(
            contribution=summary,
            previous_summary=f"Merging branch `{branch_name}` with purpose: {meta.purpose if meta else ''}",
            update_roadmap=f"Merged `{branch_name}`: {summary}",
        )

        # Mark branch as merged in metadata
        if meta:
            meta.status = "merged"
            meta.merged_into = target
            meta.merged_at = _now()
            self._write(self._meta_path(branch_name), meta.to_yaml())

        return merge_commit

    def context(self, branch: Optional[str] = None, k: int = 1) -> ContextResult:
        """
        CONTEXT command (paper S3.5).

        Retrieves historical context at varying resolutions. The paper fixes
        K=1 in experiments (only the most recent commit record is revealed).

        Parameters
        ----------
        branch : branch to retrieve context from (default: current branch)
        k      : number of recent commit records to surface (paper default K=1)
        """
        target = branch or self._current_branch
        commits = self._read_commits(target)
        ota = self._read_ota(target)
        roadmap = self._read(self._main_md())
        meta_text = self._read(self._meta_path(target))
        meta = BranchMetadata.from_yaml(meta_text) if meta_text else None

        return ContextResult(
            branch_name=target,
            k=k,
            commits=commits[-k:] if commits else [],
            ota_records=ota,
            main_roadmap=roadmap,
            metadata=meta,
        )

    @property
    def current_branch(self) -> str:
        return self._current_branch

    def switch_branch(self, name: str) -> None:
        """Switch the active branch without creating a new one."""
        if not self._branch_dir(name).exists():
            raise FileNotFoundError(f"Branch '{name}' does not exist.")
        self._current_branch = name

    def list_branches(self) -> list[str]:
        branches_root = self.gcc_dir / "branches"
        if not branches_root.exists():
            return []
        return [d.name for d in branches_root.iterdir() if d.is_dir()]

    def update_roadmap(self, content: str) -> None:
        """Overwrite or append to main.md global roadmap."""
        self._append(self._main_md(), f"\n{content}\n")

File: LUA/contexa/workspace.lua
--- contexa — Git-inspired context management for LLM agents
--- GCCWorkspace: persistent versioned memory workspace.
--- Based on arXiv:2508.00031.
---
--- @module contexa.workspace

local models = require("contexa.models")
local lfs_ok, lfs = pcall(require, "lfs")

local W = {}
W.__index = W

local MAIN_BRANCH = "main"
local GCC_DIR = ".GCC"

------------------------------------------------------------------------
-- Filesystem helpers
------------------------------------------------------------------------

local function path_join(...)
    local parts = { ... }
    return table.concat(parts, "/")
end

local function dir_exists(path)
    if lfs_ok then
        local attr = lfs.attributes(path)
        return attr and attr.mode == "directory"
    end
    -- fallback: try to open directory
    local ok, _, code = os.rename(path, path)
    if ok then return true end
    -- code 13 = permission denied (exists but can't rename)
    return code == 13
end

local function file_exists(path)
    local f = io.open(path, "r")
    if f then f:close(); return true end
    return false
end

local function mkdir_p(path)
    if lfs_ok then
        -- split and create each segment
        local accum = ""
        for seg in path:gmatch("[^/]+") do
            accum = accum == "" and seg or (accum .. "/" .. seg)
            if not dir_exists(accum) then
                lfs.mkdir(accum)
            end
        end
        return
    end
    -- fallback: use os.execute
    os.execute('mkdir -p "' .. path .. '"')
end

local function read_file(path)
    local f = io.open(path, "r")
    if not f then return "" end
    local content = f:read("*a")
    f:close()
    return content or ""
end

local function write_file(path, content)
    local f = io.open(path, "w")
    if not f then error("Cannot write to " .. path) end
    f:write(content)
    f:close()
end

local function append_file(path, content)
    local f = io.open(path, "a")
    if not f then error("Cannot append to " .. path) end
    f:write(content)
    f:close()
end

local function list_dirs(path)
    local dirs = {}
    if lfs_ok then
        for entry in lfs.dir(path) do
            if entry ~= "." and entry ~= ".." then
                local full = path_join(path, entry)
                local attr = lfs.attributes(full)
                if attr and attr.mode == "directory" then
                    dirs[#dirs + 1] = entry
                end
            end
        end
    else
        -- fallback: use ls
        local handle = io.popen('ls -1 "' .. path .. '" 2>/dev/null')
        if handle then
            for line in handle:lines() do
                if line ~= "" then
                    local full = path_join(path, line)
                    if dir_exists(full) then
                        dirs[#dirs + 1] = line
                    end
                end
            end
            handle:close()
        end
    end
    table.sort(dirs)
    return dirs
end

------------------------------------------------------------------------
-- Parsing helpers
------------------------------------------------------------------------

--- Split a string by a delimiter pattern.
--- @param text string
--- @param delim string
--- @return string[]
local function split(text, delim)
    local parts = {}
    local pos = 1
    while true do
        local s, e = text:find(delim, pos, true)  -- plain find
        if not s then
            parts[#parts + 1] = text:sub(pos)
            break
        end
        parts[#parts + 1] = text:sub(pos, s - 1)
        pos = e + 1
    end
    return parts
end

--- Extract a field value from a markdown block by prefix.
--- @param block string
--- @param prefix string  e.g. "**Timestamp:**"
--- @return string
local function extract_field(block, prefix)
    -- Escape Lua pattern special chars in prefix
    local escaped = prefix:gsub("([%(%)%.%%%+%-%*%?%[%]%^%$])", "%%%1")
    local val = block:match(escaped .. "%s*(.-)%s*\n")
    if not val then
        -- Try matching to end of string (last field in block)
        val = block:match(escaped .. "%s*(.-)%s*$")
    end
    return val or ""
end

--- Parse commit records from commit.md content.
--- @param text string
--- @return CommitRecord[]
local function parse_commits(text)
    local records = {}
    local blocks = split(text, "\n---\n")
    for _, block in ipairs(blocks) do
        block = block:match("^%s*(.-)%s*$") or "" -- trim
        if block ~= "" then
            local commit_id = block:match("## Commit `([^`]+)`") or ""
            if commit_id ~= "" then
                local rec = {
                    commit_id = commit_id,
                    timestamp = extract_field(block, "**Timestamp:**"),
                    branch_name = "",
                    branch_purpose = extract_field(block, "**Branch Purpose:**"),
                    previous_progress_summary = extract_field(block, "**Previous Progress Summary:**"),
                    this_commit_contribution = extract_field(block, "**This Commit's Contribution:**"),
                }
                records[#records + 1] = rec
            end
        end
    end
    return records
end

--- Parse OTA records from log.md content.
--- @param text string
--- @return OTARecord[]
local function parse_ota(text)
    local records = {}
    local blocks = split(text, "\n---\n")
    for _, block in ipairs(blocks) do
        block = block:match("^%s*(.-)%s*$") or "" -- trim
        if block ~= "" then
            local step_str, ts = block:match("### Step (%d+) %— (.-)%s*\n")
            if not step_str then
                step_str, ts = block:match("### Step (%d+)%-(.-)%s*\n")
            end
            if not step_str then
                step_str, ts = block:match("### Step (%d+) %— (.-)%s*$")
            end
            local step = tonumber(step_str) or 0
            local obs = extract_field(block, "**Observation:**")
            local thought = extract_field(block, "**Thought:**")
            local action = extract_field(block, "**Action:**")
            if obs ~= "" or thought ~= "" or action ~= "" then
                records[#records + 1] = {
                    step = step,
                    timestamp = ts or "",
                    observation = obs,
                    thought = thought,
                    action = action,
                }
            end
        end
    end
    return records
end

------------------------------------------------------------------------
-- GCCWorkspace
------------------------------------------------------------------------

--- Create a new GCCWorkspace.
--- @param project_root string
--- @return GCCWorkspace
function W.new(project_root)
    local self = setmetatable({}, W)
    self.root = project_root
    self.gcc_dir = path_join(project_root, GCC_DIR)
    self._current_branch = MAIN_BRANCH
    return self
end

--- Initialize a fresh .GCC workspace.
--- @param project_roadmap string|nil
function W:init(project_roadmap)
    if dir_exists(self.gcc_dir) then
        error("Workspace already exists at " .. self.gcc_dir)
    end

    local branch_dir = path_join(self.gcc_dir, "branches", MAIN_BRANCH)
    mkdir_p(branch_dir)

    local ts = models.timestamp()
    local roadmap = project_roadmap or ""
    write_file(
        path_join(self.gcc_dir, "main.md"),
        string.format("# Project Roadmap\n\n**Initialized:** %s\n\n%s\n", ts, roadmap)
    )

    write_file(
        path_join(branch_dir, "log.md"),
        "# OTA Log — branch `main`\n\n"
    )
    write_file(
        path_join(branch_dir, "commit.md"),
        "# Commit History — branch `main`\n\n"
    )

    local meta = models.new_branch_metadata(MAIN_BRANCH, "Primary reasoning trajectory", "", ts)
    write_file(
        path_join(branch_dir, "metadata.yaml"),
        models.metadata_to_yaml(meta)
    )

    self._current_branch = MAIN_BRANCH
end

--- Attach to an existing .GCC workspace.
function W:load()
    if not dir_exists(self.gcc_dir) then
        error("No workspace found at " .. self.gcc_dir)
    end
    self._current_branch = MAIN_BRANCH
end

--- Get current branch name.
--- @return string
function W:current_branch()
    return self._current_branch
end

--- Get the directory path for a branch.
--- @param name string|nil
--- @return string
function W:_branch_dir(name)
    return path_join(self.gcc_dir, "branches", name or self._current_branch)
end

--- Log an OTA cycle to the current branch.
--- @param observation string
--- @param thought string
--- @param action string
--- @return OTARecord
function W:log_ota(observation, thought, action)
    local branch_dir = self:_branch_dir()
    local log_path = path_join(branch_dir, "log.md")
    local existing = parse_ota(read_file(log_path))
    local step = #existing + 1
    local rec = models.new_ota(step, observation, thought, action)
    append_file(log_path, models.ota_to_markdown(rec))
    return rec
end

--- COMMIT — checkpoint a milestone on the current branch.
--- @param contribution string
--- @param previous_summary string|nil
--- @param update_roadmap string|nil
--- @return CommitRecord
function W:commit(contribution, previous_summary, update_roadmap)
    local branch_dir = self:_branch_dir()
    local meta_text = read_file(path_join(branch_dir, "metadata.yaml"))
    local meta = models.metadata_from_yaml(meta_text)
    local branch_purpose = meta.purpose or ""

    if not previous_summary or previous_summary == "" then
        local commits = parse_commits(read_file(path_join(branch_dir, "commit.md")))
        if #commits > 0 then
            previous_summary = commits[#commits].this_commit_contribution
        else
            previous_summary = "Initial state — no prior commits."
        end
    end

    local rec = models.new_commit(
        self._current_branch, branch_purpose,
        previous_summary, contribution
    )
    append_file(
        path_join(branch_dir, "commit.md"),
        models.commit_to_markdown(rec)
    )

    if update_roadmap and update_roadmap ~= "" then
        local ts = models.timestamp()
        append_file(
            path_join(self.gcc_dir, "main.md"),
            string.format("\n## Update (%s)\n%s\n", ts, update_roadmap)
        )
    end

    return rec
end

--- BRANCH — create an isolated reasoning workspace.
--- @param name string
--- @param purpose string
function W:branch(name, purpose)
    local branch_dir = path_join(self.gcc_dir, "branches", name)
    if dir_exists(branch_dir) then
        error("Branch already exists: " .. name)
    end

    mkdir_p(branch_dir)

    write_file(
        path_join(branch_dir, "log.md"),
        string.format("# OTA Log — branch `%s`\n\n", name)
    )
    write_file(
        path_join(branch_dir, "commit.md"),
        string.format("# Commit History — branch `%s`\n\n", name)
    )

    local meta = models.new_branch_metadata(name, purpose, self._current_branch)
    write_file(
        path_join(branch_dir, "metadata.yaml"),
        models.metadata_to_yaml(meta)
    )

    self._current_branch = name
end

--- Switch to an existing branch.
--- @param name string
function W:switch_branch(name)
    local branch_dir = path_join(self.gcc_dir, "branches", name)
    if not dir_exists(branch_dir) then
        error("Branch does not exist: " .. name)
    end
    self._current_branch = name
end

--- List all branches.
--- @return string[]
function W:list_branches()
    local branches_dir = path_join(self.gcc_dir, "branches")
    if not dir_exists(branches_dir) then return {} end
    return list_dirs(branches_dir)
end

--- MERGE — integrate a branch back into a target.
--- @param branch_name string
--- @param summary string|nil
--- @param target string|nil
--- @return CommitRecord
function W:merge(branch_name, summary, target)
    target = target or MAIN_BRANCH

    local src_dir = path_join(self.gcc_dir, "branches", branch_name)
    local tgt_dir = path_join(self.gcc_dir, "branches", target)

    if not dir_exists(src_dir) then
        error("Source branch does not exist: " .. branch_name)
    end
    if not dir_exists(tgt_dir) then
        error("Target branch does not exist: " .. target)
    end

    local src_commits = parse_commits(read_file(path_join(src_dir, "commit.md")))
    local src_ota = parse_ota(read_file(path_join(src_dir, "log.md")))
    local src_meta_text = read_file(path_join(src_dir, "metadata.yaml"))
    local src_meta = models.metadata_from_yaml(src_meta_text)

    -- Auto-generate summary if not provided
    if not summary or summary == "" then
        local contributions = {}
        for _, c in ipairs(src_commits) do
            contributions[#contributions + 1] = c.this_commit_contribution
        end
        summary = string.format(
            "Merged branch `%s` (%d commits). Contributions: %s",
            branch_name, #src_commits, table.concat(contributions, " | ")
        )
    end

    -- Append OTA records from source to target
    if #src_ota > 0 then
        local ts = models.timestamp()
        local header = string.format("\n## Merged from `%s` (%s)\n\n", branch_name, ts)
        append_file(path_join(tgt_dir, "log.md"), header)
        for _, rec in ipairs(src_ota) do
            append_file(path_join(tgt_dir, "log.md"), models.ota_to_markdown(rec))
        end
    end

    -- Switch to target and create merge commit
    self._current_branch = target
    local prev = string.format("Merging branch `%s` with purpose: %s", branch_name, src_meta.purpose or "")
    local roadmap_update = string.format("Merged `%s`: %s", branch_name, summary)
    local commit_rec = self:commit(summary, prev, roadmap_update)

    -- Update source metadata
    local ts = models.timestamp()
    src_meta.status = "merged"
    src_meta.merged_into = target
    src_meta.merged_at = ts
    write_file(path_join(src_dir, "metadata.yaml"), models.metadata_to_yaml(src_meta))

    return commit_rec
end

--- CONTEXT — hierarchical memory retrieval.
--- @param branch string|nil
--- @param k number|nil
--- @return ContextResult
function W:context(branch, k)
    branch = branch or self._current_branch
    k = k or 1

    local branch_dir = path_join(self.gcc_dir, "branches", branch)
    if not dir_exists(branch_dir) then
        error("Branch does not exist: " .. branch)
    end

    local all_commits = parse_commits(read_file(path_join(branch_dir, "commit.md")))
    local all_ota = parse_ota(read_file(path_join(branch_dir, "log.md")))
    local roadmap = read_file(path_join(self.gcc_dir, "main.md"))
    local meta_text = read_file(path_join(branch_dir, "metadata.yaml"))
    local meta = nil
    if meta_text ~= "" then
        meta = models.metadata_from_yaml(meta_text)
    end

    -- Last k commits
    local commits = {}
    local start = math.max(1, #all_commits - k + 1)
    for i = start, #all_commits do
        commits[#commits + 1] = all_commits[i]
    end

    local result = models.new_context_result(branch, k)
    result.commits = commits
    result.ota_records = all_ota
    result.main_roadmap = roadmap
    result.metadata = meta

    return result
end

return W
